include/ginkgo/core/base/batch_lin_op.hpp

```diff
@@ -0,0 +1,386 @@
+ * representable through or with the LinOp class.
+ *
+ * A batched operator is defined as a set of independent linear operators which
+ * have no communication/information exchange between them. Therefore, any
+ * collective operations between the batches is not possible and not
+ * implemented. This allows for each batch to be computed and operated on in an
+ * embarrassingly parallel fashion.
+ *
+ * A key difference between the LinOp and the BatchLinOp class is that the apply
+ * between BatchLinOps is no longer supported. The user can apply a BatchLinOp
+ * to a batch::MultiVector but not to any general BatchLinOp. Therefore, the
+ * BatchLinOp serves only as a base class providing necessary core functionality
+ * from Polymorphic object and store the dimensions of the batched object.
+ *
+ * @ref BatchLinOp
+ */
+class BatchLinOp : public EnableAbstractPolymorphicObject<BatchLinOp> {
+public:
+    /**
+     * Returns the number of items in the batch operator.
+     *
+     * @return number of items in the batch operator
+     */
+    size_type get_num_batch_items() const noexcept
+    {
+        return size_.get_num_batch_items();
+    }
+
+    /**
+     * Returns the common size of the batch items.
+     *
+     * @return the common size stored
+     */
+    dim<2> get_common_size() const { return size_.get_common_size(); }
+
+    /**
+     * Returns the size of the batch operator.
+     *
+     * @return size of the batch operator, a batch_dim object
+     */
+    const batch_dim<2>& get_size() const noexcept { return size_; }
+
+protected:
+    /**
+     * Sets the size of the batch operator.
+     *
+     * @param size to be set
+     */
+    void set_size(const batch_dim<2>& size) { size_ = size; }
+
+    /**
+     * Creates a batch operator storing items of uniform sizes.
+     *
+     * @param exec        the executor where all the operations are performed
+     * @param num_batch_items the number of batch items to be stored in the
+     * operator
+     * @param size        the common size of the items in the batched operator
+     */
+    explicit BatchLinOp(std::shared_ptr<const Executor> exec,
+                        const size_type num_batch_items = 0,
+                        const dim<2>& common_size = dim<2>{})
+        : EnableAbstractPolymorphicObject<BatchLinOp>(exec),
+          size_{num_batch_items > 0 ? batch_dim<2>(num_batch_items, common_size)
+                                    : batch_dim<2>{}}
+    {}
+
+    /**
+     * Creates a batch operator storing items of uniform sizes.
+     *
+     * @param exec        the executor where all the operations are performed
+     * @param batch_size  the size the batched operator, as a batch_dim object
+     */
+    explicit BatchLinOp(std::shared_ptr<const Executor> exec,
+                        const batch_dim<2>& batch_size)
+        : EnableAbstractPolymorphicObject<BatchLinOp>(exec), size_{batch_size}
+    {}
```

Comment:
These two constructors essentially do the same thing. If we want to have both of them, then one of them should be implemented in terms of the other, so we have less potential for inconsistency.
```suggestion
    explicit BatchLinOp(std::shared_ptr<const Executor> exec,
                        const size_type num_batch_items = 0,
                        const dim<2>& common_size = dim<2>{})
        : BatchLinOp{std::move(exec), num_batch_items > 0 ? batch_dim<2>(num_batch_items, common_size)
                                    : batch_dim<2>{}}
    {}

    /**
     * Creates a batch operator storing items of uniform sizes.
     *
     * @param exec        the executor where all the operations are performed
     * @param batch_size  the size the batched operator, as a batch_dim object
     */
    explicit BatchLinOp(std::shared_ptr<const Executor> exec,
                        const batch_dim<2>& batch_size)
        : EnableAbstractPolymorphicObject<BatchLinOp>(exec), size_{batch_size}
    {}
```